skimage/segmentation/active_contour_model.py

```diff
@@ -1,218 +1,249 @@
+                               img.T, kx=2, ky=2, s=0)
+
+    x, y = snake[:, 0].astype(np.float), snake[:, 1].astype(np.float)
+    n = len(x)
+    xsave = np.empty((convergence_order, n))
+    ysave = np.empty((convergence_order, n))
+
+    # Build snake shape matrix for Euler equation
+    a = np.roll(np.eye(n), -1, axis=0) + \
+        np.roll(np.eye(n), -1, axis=1) - \
+        2*np.eye(n)  # second order derivative, central difference
+    b = np.roll(np.eye(n), -2, axis=0) + \
+        np.roll(np.eye(n), -2, axis=1) - \
+        4*np.roll(np.eye(n), -1, axis=0) - \
+        4*np.roll(np.eye(n), -1, axis=1) + \
+        6*np.eye(n)  # fourth order derivative, central difference
+    A = -alpha*a + beta*b
+
+    # Impose boundary conditions different from periodic:
+    sfixed = False
+    if boundary_condition.startswith('fixed'):
+        A[0, :] = 0
+        A[1, :] = 0
+        A[1, :3] = [1, -2, 1]
+        sfixed = True
+    efixed = False
+    if boundary_condition.endswith('fixed'):
+        A[-1, :] = 0
+        A[-2, :] = 0
+        A[-2, -3:] = [1, -2, 1]
+        efixed = True
+    sfree = False
+    if boundary_condition.startswith('free'):
+        A[0, :] = 0
+        A[0, :3] = [1, -2, 1]
+        A[1, :] = 0
+        A[1, :4] = [-1, 3, -3, 1]
+        sfree = True
+    efree = False
+    if boundary_condition.endswith('free'):
+        A[-1, :] = 0
+        A[-1, -3:] = [1, -2, 1]
+        A[-2, :] = 0
+        A[-2, -4:] = [-1, 3, -3, 1]
+        efree = True
+
+    # Only one inversion is needed for implicit spline energy minimization:
+    inv = np.linalg.inv(A + gamma*np.eye(n))
+
+    # Explicit time stepping for image energy minimization:
+    for i in range(max_iterations):
+        fx = intp(x, y, dx=1, grid=False)
+        fy = intp(x, y, dy=1, grid=False)
+        if sfixed:
+            fx[0] = 0
+            fy[0] = 0
+        if efixed:
+            fx[-1] = 0
+            fy[-1] = 0
+        if sfree:
+            fx[0] *= 2
+            fy[0] *= 2
+        if efree:
+            fx[-1] *= 2
+            fy[-1] *= 2
+        xn = inv @ (gamma*x + fx)
+        yn = inv @ (gamma*y + fy)
+
+        # Movements are capped to max_px_move per iteration:
+        dx = max_px_move*np.tanh(xn-x)
+        dy = max_px_move*np.tanh(yn-y)
+        if sfixed:
+            dx[0] = 0
+            dy[0] = 0
+        if efixed:
+            dx[-1] = 0
+            dy[-1] = 0
+        x += dx
+        y += dy
+
+        # Convergence criteria needs to compare to a number of previous
+        # configurations since oscillations can occur.
+        j = i % (convergence_order+1)
+        if j < convergence_order:
+            xsave[j, :] = x
+            ysave[j, :] = y
+        else:
+            dist = np.min(np.max(np.abs(xsave-x[None, :]) +
+                                 np.abs(ysave-y[None, :]), 1))
+            if dist < convergence:
+                break
+
+    if coordinates == 'xy':
+        x, y = y, x
```

Comment:
This is so confusion, so were you using rc coordinates, or xy coordinates in this codebase?

Reply:
The codebase was xy, but I swapped them at the top with `snake = snake[:, ::-1]`. I agree it's confusing, which is why I'm deprecating it. =P

Reply:
Apologies for the large diff: the file was incorrectly encoded as windows (with `^M` everywhere), so I updated it. Locally, git can ignore these differences, but apparently GitHub was not smart enough to use that flag. I recommend checking out the diff locally with `git diff --ignore-cr-at-eol`.

Yes, it's ironic that it's `cr`. =P

Reply:
I think you should rename the variables internal to this function too.

Otherwise, maybe it means something that `x,y` should be kept ;).

Reply:
I wanted to keep the diff small. I know GitHub doesn't show it but I promise, it's pretty small. =)

Reply:
![image](https://user-images.githubusercontent.com/90008/62911489-8d790f80-bd52-11e9-8d71-391c487d64bc.png)

Reply:
I guess I want to say please consider this diff

```patch
diff --git a/skimage/segmentation/active_contour_model.py b/skimage/segmentation/active_contour_model.py
index dc10fcb07..165ffcb83 100644
--- a/skimage/segmentation/active_contour_model.py
+++ b/skimage/segmentation/active_contour_model.py
@@ -112,8 +112,9 @@ def active_contour(image, snake, alpha=0.01, beta=0.1,
                    '0.18, but will stop working thereafter.')
         warn(message, category=FutureWarning, stacklevel=2)
         coordinates = 'xy'
+        snake_xy = snake
     if coordinates == 'rc':
-        snake = snake[:, ::-1]
+        snake_xy = snake[:, ::-1]
     max_iterations = int(max_iterations)
     if max_iterations <= 0:
         raise ValueError("max_iterations should be >0.")
@@ -153,7 +154,7 @@ def active_contour(image, snake, alpha=0.01, beta=0.1,
                                np.arange(img.shape[0]),
                                img.T, kx=2, ky=2, s=0)
 
-    x, y = snake[:, 0].astype(np.float), snake[:, 1].astype(np.float)
+    x, y = snake_xy[:, 0].astype(np.float), snake_xy[:, 1].astype(np.float)
     n = len(x)
     xsave = np.empty((convergence_order, n))
     ysave = np.empty((convergence_order, n))
@@ -244,6 +245,6 @@ def active_contour(image, snake, alpha=0.01, beta=0.1,
                 break
 
     if coordinates == 'xy':
-        x, y = y, x
-
-    return np.stack([y, x], axis=1)
+        return np.stack([x, y], axis=1)
+    else:
+        return np.stack([y, x], axis=1)
```

This is code style, but I find that when the units/coordinates are ambiguous in the code, it is nice to be explicit about the units. Flipping the meaning of x and y at the very end really irks me.